**Enrique Santos Fraire, 09/09/2022**

# **CARGA Y LIMPIEZA DE DATOS**

In [ ]:
from google.colab import drive

drive.mount("/content/gdrive")  
!pwd  # show current path 

Mounted at /content/gdrive
/content


In [ ]:
%cd "/content/gdrive/MyDrive/Semestre IA/Estadística para la ciencia de datos"
!ls  # show current directory

/content/gdrive/MyDrive/Semestre IA/Estadística para la ciencia de datos
' Business case: Understanding profitability in the US financial market.ipynb'
 us2022q2a.csv
 usfirms2022.csv
'Workshop 1, Advanced AI - Statistics Module.ipynb'
'Workshop 2, Advanced AI - Statistics Module.ipynb'
'Workshop 3, Advanced AI - Statistics Module.ipynb'


In [ ]:
import pandas as pd
import plotly.express as px

df1 = pd.read_csv('us2022q2a.csv')
df2 = pd.read_csv('usfirms2022.csv')
usdata = df1.copy()

In [ ]:
usdata['qdate'] = pd.PeriodIndex(usdata['q'], freq="Q")

usdata.set_index(['firm', 'qdate'], inplace = True)

usdata.index
usdata.columns
usdata['lprice'] = usdata.groupby(['firm'])['adjprice'].shift(4)

In [ ]:
usdata[['q', 'adjprice', 'lprice']]
#usdata.loc['ADT': 'ADTN', ['adjprice', 'lprice']].iloc[84:98,]

q   adjprice    lprice
firm qdate                              
A    2000Q1  2000q1  63.761670       NaN
     2000Q2  2000q2  45.215607       NaN
     2000Q3  2000q3  30.003238       NaN
     2000Q4  2000q4  33.566841       NaN
     2001Q1  2001q1  18.840347  63.76167
...             ...        ...       ...
ZYNE 2021Q3  2021q3   4.240000   3.31000
     2021Q4  2021q4   2.880000   3.30000
     2022Q1  2022q1   2.050000   4.65000
     2022Q2  2022q2   1.140000   5.29000
NaN  NaN        NaN        NaN       NaN

[324447 rows x 3 columns]

In [ ]:
import numpy as np
usdata['r'] = np.log(usdata['adjprice']) - np.log(usdata['lprice'])
usdata['r2']= np.log(usdata['adjprice']) - np.log(usdata.groupby(['firm'])['adjprice'].shift(4))

In [ ]:
usdata[['q', 'adjprice', 'lprice', 'r', 'r2']]
#usdata.loc['ADT': 'ADTN', ['adjprice', 'lprice', 'r', 'r2']].iloc[85:96,]

q   adjprice    lprice         r        r2
firm qdate                                                  
A    2000Q1  2000q1  63.761670       NaN       NaN       NaN
     2000Q2  2000q2  45.215607       NaN       NaN       NaN
     2000Q3  2000q3  30.003238       NaN       NaN       NaN
     2000Q4  2000q4  33.566841       NaN       NaN       NaN
     2001Q1  2001q1  18.840347  63.76167 -1.219152 -1.219152
...             ...        ...       ...       ...       ...
ZYNE 2021Q3  2021q3   4.240000   3.31000  0.247615  0.247615
     2021Q4  2021q4   2.880000   3.30000 -0.136132 -0.136132
     2022Q1  2022q1   2.050000   4.65000 -0.819027 -0.819027
     2022Q2  2022q2   1.140000   5.29000 -1.534790 -1.534790
NaN  NaN        NaN        NaN       NaN       NaN       NaN

[324447 rows x 5 columns]

OBTENEMOS EL ÚLTIMO PRECIO ANUAL Y LOS RETORNOS COMPUESTOS

In [ ]:
df1['lprice'] = df1.groupby(['firm'])['adjprice'].shift(4)
df1['r'] = np.log(df1['adjprice']) - np.log(df1['lprice'])
df1['r2']= np.log(df1['adjprice']) - np.log(df1.groupby(['firm'])['adjprice'].shift(4))
df1.head(10)

firm       q    revenue       cogs       sgae  otheropexp  extraincome  \
0    A  2000q1        NaN        NaN        NaN         NaN          NaN   
1    A  2000q2  2485000.0  1261000.0  1010000.0         0.0      42000.0   
2    A  2000q3  2670000.0  1369000.0  1091000.0         0.0      28000.0   
3    A  2000q4  3372000.0  1732000.0  1182000.0         0.0      10000.0   
4    A  2001q1  2841000.0  1449000.0  1113000.0         0.0      -6000.0   
5    A  2001q2  2382000.0  1406000.0  1038000.0         0.0     273000.0   
6    A  2001q3  1806000.0  1146000.0  1002000.0         0.0      30000.0   
7    A  2001q4  1660000.0  1337000.0   981000.0         0.0     588000.0   
8    A  2002q1  1426000.0   878000.0   990000.0         0.0      21000.0   
9    A  2002q2  1457000.0   854000.0   952000.0         0.0      16000.0   

   finexp  incometax  totalassets  ...  stockholderequity   adjprice  \
0     NaN        NaN          NaN  ...                NaN  63.761670   
1     0.0    90000.0    7321000.0  ...          4642000.0  45.215607   
2     0.0    83000.0    7827000.0  ...          4902000.0  30.003238   
3     0.0   163000.0    8425000.0  ...          5265000.0  33.566841   
4     0.0   119000.0    9208000.0  ...          5541000.0  18.840347   
5     0.0   115000.0    9080000.0  ...          5616000.0  19.925522   
6     0.0   -93000.0    8622000.0  ...          5487000.0  11.985967   
7     0.0  -213000.0    7986000.0  ...          5659000.0  17.479281   
8     0.0  -106000.0    8558000.0  ...          5395000.0  21.433730   
9     0.0   -80000.0    8238000.0  ...          5128000.0  14.499649   

   originalprice  sharesoutstanding  fiscalmonth    year  cto     lprice  \
0     104.000000         452000.000          NaN  2000.0  1.0        NaN   
1      73.750000         452271.967          6.0  2000.0  2.0        NaN   
2      48.937500         453014.579          9.0  2000.0  3.0        NaN   
3      54.750000         456366.381         12.0  2000.0  4.0        NaN   
4      30.730000         456769.737          3.0  2001.0  1.0  63.761670   
5      32.500000         457072.749          6.0  2001.0  2.0  45.215607   
6      19.549999         460637.309          9.0  2001.0  3.0  30.003238   
7      28.510000         463695.160         12.0  2001.0  4.0  33.566841   
8      34.959999         463859.978          3.0  2002.0  1.0  18.840347   
9      23.650000         464247.548          6.0  2002.0  2.0  19.925522   

          r        r2  
0       NaN       NaN  
1       NaN       NaN  
2       NaN       NaN  
3       NaN       NaN  
4 -1.219152 -1.219152  
5 -0.819441 -0.819441  
6 -0.917569 -0.917569  
7 -0.652522 -0.652522  
8  0.128965  0.128965  
9 -0.317877 -0.317877  

[10 rows x 23 columns]

In [ ]:
df2.head()

Ticker                   Name  N  Class Country\nof Origin Type of Asset  \
0   FLWS  1 800 Flowers Com Inc  1  Com A                 US         Stock   
1    TXG      10x Genomics, Inc  2  Com A                 US         Stock   
2   GOED      1847 Goedeker Inc  3    Com                 US         Stock   
3   ONEM  1life Healthcare, Inc  4    Com                 US         Stock   
4   SRCE        1st Source Corp  5    Com                 US         Stock   

               Sector NAICS\nlevel 1 Exchange / Src    Sector\nEconomatica  \
0                       Retail Trade         NASDAQ                  Trade   
1                      Manufacturing         NASDAQ      Electric Electron   
2                       Retail Trade           AMEX                  Trade   
3  Health Care and Social Assistance         NASDAQ                  Other   
4              Finance and Insurance         NASDAQ  Finance and Insurance   

                        Sector NAICS\nlast available partind  
0                                       Retail Trade       -  
1  Navigational, Measuring, Electromedical, and C...       -  
2              Furniture and Home Furnishings Stores       -  
3                              Offices of Physicians       -  
4                   Depository Credit Intermediation       -

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324447 entries, 0 to 324446
Data columns (total 23 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   firm               324446 non-null  object 
 1   q                  324446 non-null  object 
 2   revenue            158614 non-null  float64
 3   cogs               158309 non-null  float64
 4   sgae               158309 non-null  float64
 5   otheropexp         158311 non-null  float64
 6   extraincome        158613 non-null  float64
 7   finexp             158614 non-null  float64
 8   incometax          158614 non-null  float64
 9   totalassets        159808 non-null  float64
 10  totalliabilities   159808 non-null  float64
 11  shortdebt          129677 non-null  float64
 12  longdebt           159595 non-null  float64
 13  stockholderequity  159740 non-null  float64
 14  adjprice           159154 non-null  float64
 15  originalprice      159234 non-null  float64
 16  sh

In [ ]:
df1.isna().sum()

firm                      1
q                         1
revenue              165833
cogs                 166138
sgae                 166138
otheropexp           166136
extraincome          165834
finexp               165833
incometax            165833
totalassets          164639
totalliabilities     164639
shortdebt            194770
longdebt             164852
stockholderequity    164707
adjprice             165293
originalprice        165213
sharesoutstanding    166625
fiscalmonth          164633
year                      1
cto                       1
lprice               179561
r                    179670
r2                   179670
dtype: int64

DROPEAMOS LOS DATOS QUE NO APORTAN INFORMACIÓN AL MODELO, FISCALMONTH, YEAR Y CTO YA ESTAN EN Q. COUNTRY OF ORIGIN Y ASSET ES INFORMACIÓN QUE NO CAMBIA.

In [ ]:
df1 = df1.drop(['fiscalmonth', 'year', 'cto'], axis=1)
df2 = df2.drop(['N', 'Country\nof Origin', 'Type of Asset'], axis=1)

UNIMOS AMBOS DATA FRAMES.

In [ ]:
df_merge = df1.merge(df2, left_on='firm', right_on='Ticker')
df_merge

firm       q    revenue       cogs          sgae  otheropexp  \
0          A  2000q1        NaN        NaN           NaN         NaN   
1          A  2000q2  2485000.0  1261000.0  1.010000e+06         0.0   
2          A  2000q3  2670000.0  1369000.0  1.091000e+06         0.0   
3          A  2000q4  3372000.0  1732000.0  1.182000e+06         0.0   
4          A  2001q1  2841000.0  1449000.0  1.113000e+06         0.0   
...      ...     ...        ...        ...           ...         ...   
323811  ZYNE  2021q2        0.0        0.0  9.838494e+03         0.0   
323812  ZYNE  2021q3        0.0        0.0  1.021065e+04         0.0   
323813  ZYNE  2021q4        0.0        0.0  8.836436e+03         0.0   
323814  ZYNE  2022q1        0.0        0.0  8.903915e+03         0.0   
323815  ZYNE  2022q2        0.0        0.0  9.168770e+03         0.0   

         extraincome  finexp  incometax  totalassets  ...         r        r2  \
0                NaN     NaN        NaN          NaN  ...       NaN       NaN   
1       42000.000000   0.000    90000.0  7321000.000  ...       NaN       NaN   
2       28000.000000   0.000    83000.0  7827000.000  ...       NaN       NaN   
3       10000.000000   0.000   163000.0  8425000.000  ...       NaN       NaN   
4       -6000.000000   0.000   119000.0  9208000.000  ... -1.219152 -1.219152   
...              ...     ...        ...          ...  ...       ...       ...   
323811   -117.528220  -5.943        0.0    98195.904  ...  0.450906  0.450906   
323812   -376.636750  -5.038        0.0    89996.170  ...  0.247615  0.247615   
323813     16.937906  -4.433        0.0    81171.507  ... -0.136132 -0.136132   
323814    317.252110 -96.044        0.0    74381.029  ... -0.819027 -0.819027   
323815   -775.927860 -91.691        0.0    67006.959  ... -1.534790 -1.534790   

        Ticker                          Name  Class  Sector NAICS\nlevel 1  \
0            A     Agilent Technologies, Inc    Com          Manufacturing   
1            A     Agilent Technologies, Inc    Com          Manufacturing   
2            A     Agilent Technologies, Inc    Com          Manufacturing   
3            A     Agilent Technologies, Inc    Com          Manufacturing   
4            A     Agilent Technologies, Inc    Com          Manufacturing   
...        ...                           ...    ...                    ...   
323811    ZYNE  Zynerba Pharmaceuticals, Inc    Com          Manufacturing   
323812    ZYNE  Zynerba Pharmaceuticals, Inc    Com          Manufacturing   
323813    ZYNE  Zynerba Pharmaceuticals, Inc    Com          Manufacturing   
323814    ZYNE  Zynerba Pharmaceuticals, Inc    Com          Manufacturing   
323815    ZYNE  Zynerba Pharmaceuticals, Inc    Com          Manufacturing   

        Exchange / Src  Sector\nEconomatica  \
0                 NYSE    Electric Electron   
1                 NYSE    Electric Electron   
2                 NYSE    Electric Electron   
3                 NYSE    Electric Electron   
4                 NYSE    Electric Electron   
...                ...                  ...   
323811          NASDAQ             Chemical   
323812          NASDAQ             Chemical   
323813          NASDAQ             Chemical   
323814          NASDAQ             Chemical   
323815          NASDAQ             Chemical   

                             Sector NAICS\nlast available  partind  
0       Navigational, Measuring, Electromedical, and C...    0.124  
1       Navigational, Measuring, Electromedical, and C...    0.124  
2       Navigational, Measuring, Electromedical, and C...    0.124  
3       Navigational, Measuring, Electromedical, and C...    0.124  
4       Navigational, Measuring, Electromedical, and C...    0.124  
...                                                   ...      ...  
323811          Pharmaceutical and Medicine Manufacturing        -  
323812          Pharmaceutical and Medicine Manufacturing        -  
323813          Pharmaceutical and Medicine Man

In [ ]:
df_merge.isna().sum()

firm                                 0
q                                    0
revenue                         165684
cogs                            165989
sgae                            165989
otheropexp                      165987
extraincome                     165685
finexp                          165684
incometax                       165684
totalassets                     164491
totalliabilities                164491
shortdebt                       194532
longdebt                        164704
stockholderequity               164559
adjprice                        165085
originalprice                   165005
sharesoutstanding               166480
lprice                          179329
r                               179434
r2                              179434
Ticker                               0
Name                                 0
Class                                0
Sector NAICS\nlevel 1                0
Exchange / Src                       0
Sector\nEconomatica      

In [ ]:
df_copy = df_merge.copy()

# DATOS ESTADÍSTICOS

SE CREAN LAS COLUMNAS DE DATOS ESTADÍSTICOS QUE NOS AYUDARÁN A RESOLVER LOS PROBLEMAS.

In [ ]:
# Firm Size Measures
df_copy['Book'] = df_copy['totalassets'] - df_copy['totalliabilities']
df_copy['Market'] = df_copy['originalprice'] * df_copy['sharesoutstanding']

# Profit Margin Measures
df_copy['EBIT'] = df_copy['revenue'] - df_copy['cogs'] - df_copy['sgae'] - df_copy['otheropexp']
df_copy['Net Income'] = df_copy['EBIT'] + df_copy['extraincome']- df_copy['incometax'] - df_copy['finexp']
df_copy['Earnings Per Share (EPS)'] = df_copy['Net Income'] / df_copy['sharesoutstanding']

# Se eligen los 3 márgenes para el análisis estadístico
df_copy['EPSP (deflated by price)'] = df_copy['Earnings Per Share (EPS)'] / df_copy['originalprice']

df_copy['Sales Anual Growth'] = (df_copy['revenue'] / (df_copy.groupby(['firm'])['revenue'].shift(4))) - 1
df_copy['Operating Profit Margin (OPM)'] = df_copy['EBIT'] / df_copy['revenue']
df_copy['Book-to-Market ratio'] = df_copy['Book'] / df_copy['Market']

In [ ]:
# Visualización de datos estadísticos
df_copy[['firm','q', 'Book', 'Market','EBIT','revenue','Net Income','Earnings Per Share (EPS)','EPSP (deflated by price)','Sales Anual Growth', 'Operating Profit Margin (OPM)', 'Book-to-Market ratio']].head(16)

firm       q       Book        Market      EBIT    revenue  Net Income  \
0     A  2000q1        NaN  4.700800e+07       NaN        NaN         NaN   
1     A  2000q2  4642000.0  3.335506e+07  214000.0  2485000.0    166000.0   
2     A  2000q3  4902000.0  2.216940e+07  210000.0  2670000.0    155000.0   
3     A  2000q4  5265000.0  2.498606e+07  458000.0  3372000.0    305000.0   
4     A  2001q1  5541000.0  1.403653e+07  279000.0  2841000.0    154000.0   
5     A  2001q2  5616000.0  1.485486e+07  -62000.0  2382000.0     96000.0   
6     A  2001q3  5487000.0  9.005459e+06 -342000.0  1806000.0   -219000.0   
7     A  2001q4  5659000.0  1.321995e+07 -658000.0  1660000.0    143000.0   
8     A  2002q1  5395000.0  1.621654e+07 -442000.0  1426000.0   -315000.0   
9     A  2002q2  5128000.0  1.097945e+07 -349000.0  1457000.0   -253000.0   
10    A  2002q3  5001000.0  6.095179e+06 -429000.0  1391000.0   -228000.0   
11    A  2002q4  4627000.0  8.387759e+06 -387000.0  1736000.0   -236000.0   
12    A  2003q1  4349000.0  6.196841e+06 -256000.0  1412000.0   -369000.0   
13    A  2003q2  4201000.0  9.213996e+06 -335000.0  1467000.0   -146000.0   
14    A  2003q3  2720000.0  1.051347e+07 -190000.0  1502000.0  -1556000.0   
15    A  2003q4  2824000.0  1.392260e+07   56000.0  1675000.0     13000.0   

    Earnings Per Share (EPS)  EPSP (deflated by price)  Sales Anual Growth  \
0                        NaN                       NaN                 NaN   
1                   0.367036                  0.004977                 NaN   
2                   0.342152                  0.006992                 NaN   
3                   0.668323                  0.012207                 NaN   
4                   0.337150                  0.010971                 NaN   
5                   0.210032                  0.006463           -0.041449   
6                  -0.475428                 -0.024319           -0.323596   
7                   0.308392                  0.010817           -0.507711   
8                  -0.679084                 -0.019425           -0.498064   
9                  -0.544968                 -0.023043           -0.388329   
10                 -0.488530                 -0.037407           -0.229790   
11                 -0.505327                 -0.028136            0.045783   
12                 -0.783036                 -0.059546           -0.009818   
13                 -0.309779                 -0.015845            0.006863   
14                 -3.272292                 -0.148001            0.079799   
15                  0.027302                  0.000934           -0.035138   

    Operating Profit Margin (OPM)  Book-to-Market ratio  
0                             NaN                   NaN  
1                        0.086117              0.139169  
2                        0.078652              0.221116  
3                        0.135824              0.210718  
4                        0.098205              0.394756  
5                       -0.026029              0.378058  
6                       -0.189369              0.609297  
7                       -0.396386              0.428065  
8                       -0.309958              0.332685  
9                       -0.239533              0.467054  
10                      -0.308411              0.820484  
11                      -0.222926              0.551637  
12                      -0.181303              0.701809  
13                      -0.228357              0.455937  
14                      -0.126498              0.258716  
15                       0.033433              0.202836

# **Specific questions**

# 2.2.1.1 Considering the most recent financial quarter of the dataset

## Show how many firms by industry there are in the sample

In [ ]:
mask = df_copy['q'] == '2022q2'
last_quarter = df_copy[mask]

In [ ]:
last_quarter

firm       q     revenue       cogs          sgae  otheropexp  \
89         A  2022q2   1607000.0   746000.0  5.010000e+05         0.0   
179       AA  2022q2   3644000.0  2767000.0  2.200000e+05    -75000.0   
269     AAIC  2022q2     10900.0     6374.0  0.000000e+00         0.0   
359      AAL  2022q2  13422000.0        0.0  1.240500e+07         0.0   
449     AAME  2022q2     44669.0        0.0  4.678400e+04         0.0   
...      ...     ...         ...        ...           ...         ...   
323455  ZVIA  2022q2     45542.0    28168.0  2.407400e+04      8043.0   
323545   ZVO  2022q2     51380.0    34995.0  2.610800e+04     -8882.0   
323635   ZWS  2022q2    284200.0   170400.0  6.000000e+04       300.0   
323725    ZY  2022q2      2634.0     9732.0  5.863800e+04     40460.0   
323815  ZYNE  2022q2         0.0        0.0  9.168770e+03         0.0   

        extraincome      finexp  incometax   totalassets  ...  partind  \
89      -7000.00000   20000.000    59000.0  1.045500e+07  ...    0.124   
179     81000.00000   30000.000   234000.0  1.570900e+07  ...        -   
269     -3417.00000       0.000      802.0  1.084755e+06  ...        -   
359     25000.00000  439000.000   127000.0  6.796300e+07  ...    0.032   
449         0.00000       0.000     -436.0  3.792740e+05  ...        -   
...             ...         ...        ...           ...  ...      ...   
323455   3662.00000       0.000        9.0  1.127380e+05  ...        -   
323545  -3824.00000       0.000        8.0  8.543300e+04  ...        -   
323635   -600.00000    5200.000    11300.0  1.176300e+06  ...        -   
323725   -885.00000    9376.000       11.0  4.709680e+05  ...        -   
323815   -775.92786     -91.691        0.0  6.700696e+04  ...        -   

               Book        Market          EBIT    Net Income  \
89      5122000.000  3.547756e+07  3.600000e+05  274000.00000   
179     7292000.000  8.407171e+06  7.320000e+05  549000.00000   
269      213698.000  1.138033e+05  4.526000e+03     307.00000   
359    -8422000.000  8.235848e+06  1.017000e+06  476000.00000   
449      109101.000  5.446399e+04 -2.115000e+03   -1679.00000   
...             ...           ...           ...           ...   
323455    87199.000  1.888586e+05 -1.474300e+04  -11090.00000   
323545     6605.000  3.307650e+04 -8.410000e+02   -4673.00000   
323635   229400.000  3.431356e+06  5.350000e+04   36400.00000   
323725   236623.000  1.268631e+05 -1.061960e+05 -116468.00000   
323815    57241.873  4.969939e+04 -9.168770e+03   -9853.00636   

        Earnings Per Share (EPS)  EPSP (deflated by price)  \
89                      0.917284                  0.007723   
179                     2.976438                  0.065301   
269                     0.008767                  0.002698   
359                     0.732855                  0.057796   
449                    -0.082310                 -0.030828   
...                          ...                       ...   
323455                 -0.165007                 -0.058721   
323545                 -0.137054                 -0.141279   
323635                  0.288963                  0.010608   
323725                 -1.129214                 -0.918060   
323815                 -0.226007                 -0.198252   

        Sales Anual Growth  Operating Profit Margin (OPM)  \
89                0.053770                       0.224020   
179               0.286269                       0.200878   
269               0.547197                       0.415229   
359               0.794865                       0.075771   
449              -0.131916                      -0.047348   
...                    ...                            ...   
323455            0.325745                      -0.323723   
323545           -0.257364                      -0.016368   
323635           -0.499912                       0.188248   
323725           -0.552573                     -40.317388   
323815                 NaN                   

SE OBTIENE LA CANTIDAD DE FIRMAS QUE HAY POR INDUSTRIA.

In [ ]:
firms_by_industry = df2['Sector NAICS\nlevel 1'].value_counts()
firms_by_industry

Manufacturing                                                               1567
Finance and Insurance                                                        703
Information                                                                  263
Retail Trade                                                                 152
Professional, Scientific, and Technical Services                             145
Administrative and Support and Waste Management and Remediation Services     133
Mining, Quarrying, and Oil and Gas Extraction                                104
Wholesale Trade                                                               79
Utilities                                                                     77
Transportation and Warehousing                                                69
Accommodation and Food Services                                               69
Real Estate and Rental and Leasing                                            68
Health Care and Social Assis

## For each industry (and for all industries), what can you say about the typical firm size in terms of market value and book value? How much these variables change within each industry? How firm size (in market value) is distributed?

In [ ]:
typical_firm = last_quarter.groupby('Sector NAICS\nlevel 1')['Book'].mean().to_frame()
typical_firm['Book Median'] = last_quarter.groupby('Sector NAICS\nlevel 1')['Book'].median()
typical_firm['Market Mean'] = last_quarter.groupby('Sector NAICS\nlevel 1')['Market'].mean()
typical_firm['Market Median'] = last_quarter.groupby('Sector NAICS\nlevel 1')['Market'].median()
typical_firm['OPM Mean'] = last_quarter.groupby('Sector NAICS\nlevel 1')['Operating Profit Margin (OPM)'].mean()
typical_firm['OPM Median'] = last_quarter.groupby('Sector NAICS\nlevel 1')['Operating Profit Margin (OPM)'].median()
typical_firm['Firms'] = last_quarter['Sector NAICS\nlevel 1'].value_counts()
typical_firm.reset_index(inplace=True)
typical_firm

Sector NAICS\nlevel 1          Book  \
0                                                   -  5.704446e+06   
1                     Accommodation and Food Services  5.139213e+05   
2   Administrative and Support and Waste Managemen...  2.819477e+06   
3          Agriculture, Forestry, Fishing and Hunting  3.629560e+06   
4                 Arts, Entertainment, and Recreation  5.394410e+06   
5                                        Construction  2.535521e+06   
6                                Educational Services  8.931767e+05   
7                               Finance and Insurance  5.482677e+06   
8                   Health Care and Social Assistance  1.080145e+06   
9                                         Information  4.213150e+06   
10                                      Manufacturing  2.417298e+06   
11      Mining, Quarrying, and Oil and Gas Extraction  3.345942e+06   
12      Other Services (except Public Administration)  5.954443e+05   
13   Professional, Scientific, and Technical Services  3.566048e+06   
14                 Real Estate and Rental and Leasing  1.816059e+06   
15                                       Retail Trade  3.379848e+06   
16                     Transportation and Warehousing  3.536723e+06   
17                                          Utilities  7.148295e+06   
18                                    Wholesale Trade  1.529928e+06   

    Book Median   Market Mean  Market Median  OPM Mean  OPM Median  Firms  
0   5704446.000  4.865183e+06   4.865183e+06  0.463653    0.463653      2  
1    243717.500  8.681070e+06   1.394617e+06  0.103713    0.114633     69  
2    566167.000  1.385569e+07   1.938284e+06 -0.109505    0.076840    133  
3   1104345.000  8.046780e+06   1.264045e+06 -0.016727    0.055316     16  
4     67242.769  1.278260e+07   2.504698e+06  0.038509    0.152064     21  
5    998146.500  3.857422e+06   1.745045e+06  0.106393    0.100588     45  
6    649699.000  1.302581e+06   1.524843e+06 -0.001653    0.039924     14  
7   1049158.000  8.412277e+06   1.264517e+06       NaN    0.565211    701  
8    451385.000  3.507730e+06   1.338427e+06       NaN   -0.031415     64  
9    500953.500  1.918280e+07   2.586582e+06       NaN   -0.044653    261  
10   244642.000  1.040643e+07   5.911289e+05       NaN   -0.064151   1565  
11   741145.000  6.783299e+06   1.042405e+06       NaN    0.291527    103  
12   431667.000  2.136156e+06   8.540956e+05  0.043354    0.059704     16  
13   279188.000  1.723014e+07   9.419611e+05       NaN    0.005180    145  
14   634398.000  3.625235e+06   8.819927e+05       NaN    0.145117     68  
15   577426.500  1.998810e+07   1.444971e+06 -0.006559    0.043393    152  
16  1288121.500  1.191248e+07   2.464494e+06  0.010427    0.099848     69  
17  2791950.000  1.588301e+07   6.040649e+06       NaN    0.162881     77  
18   683900.000  4.871519e+06   1.424602e+06  0.013094    0.064806     79

SE PUEDE OBSEVAR COMPARAR LA MEDIA Y LA MEDIANA TANTO DEL BOOK COMO DEL MARKET, A SIMPLE VISTA SE PUEDE VER UNA GRAN DIFERENCIA ENTRE AMBAS VARIABLES. DADO QUE EN AMBOS CASOS LA MEDIA ES MUCHO MAYOR QUE LA MEDIANA, PODEMOS INFERIR QUE HAY POCAS EMPRESAS CON VALORES MUY ALTOS QUE AFECTAN LA MEDIA, POR LO QUE ESTA DEJA DE SER REPRESENTATIVA, SIENDO LA MEJOR OPCIÓN LA MEDIANA.

In [ ]:
last_quarter['Market'].describe()

count    3.548000e+03
mean     1.095980e+07
std      6.632062e+07
min      3.490000e+01
25%      1.907523e+05
50%      1.105076e+06
75%      4.607257e+06
max      2.212838e+09
Name: Market, dtype: float64

In [ ]:
px.box(last_quarter, x = 'Sector NAICS\nlevel 1', y = 'Market')

TANTO EN LA DESCRIPCIÓN DE MARKET COMO EN EL GRÁFICO SE OBSERVA QUE AL MENOS EL 75% DE LAS EMPRESAS NO PASAN DE LOS 0.5 BILLONES, MIENTRAS QUE HAY SOLO UNAS POCAS QUE SOBREPASAN EL BILLÓN O INCLUSO LLEGAN A LOS 2 BILLONES, COMO LO ES EN LA INDUSTRIA DE MANUFACTURING, RETAIL TRADE, PROFESSIONAL, SCIENTIFIC AND TECHNICAL SERVICES E INFORMATION, QUE ES DONDE SE VE QUE AFECTADA LA MEDIA DEL MARKET.

In [ ]:
px.histogram(typical_firm, x = 'Sector NAICS\nlevel 1', y = 'Market Mean')

EN CUANTO A LA DISTRIBUCIÓN DE LA VARIABLE MARKET, VEMOS QUE SE DISTRIBUYE PRINCIPALMENTE EN LA INDUSTRIA DE INFORMATION, RETAIL TRADE, PROFESSIONAL, SCIENTIFIC AND TECHNICAL SERVICES Y UTILITIES. QUE, COMO BIEN VIMOS ANTERIORMENTE, LA MAYORÍA SON LAS QUE CUENTAN CON LAS EMPRESAS DE MÁS ALTO VALOR Y QUE INFLUYEN EN NUESTRA MEDIA.

In [ ]:
px.histogram(typical_firm, x = 'Sector NAICS\nlevel 1', y = 'Market Median')

AHORA EVALUANDO VALORES REPRESENTATIVOS COMO LA MEDIANA, SE PUEDE VER UNA DISTRIBUCIÓN MÁS NORMALIZADA, DONDE DESTACAN LAS UTILITIES, POR LO QUE EN TÉRMINOS GENERALES TIENE UNA MAYOR PRESENCIA EN EL MERCADO.

In [ ]:
last_quarter['Book'].describe()

count    3.362000e+03
mean     3.331749e+06
std      1.386098e+07
min     -1.479100e+07
25%      1.060878e+05
50%      4.577370e+05
75%      1.792766e+06
max      2.861430e+08
Name: Book, dtype: float64

In [ ]:
px.box(last_quarter, x = 'Sector NAICS\nlevel 1', y = 'Book')

DE IGUAL MANERA PODEMOS VER COMO EN BOOK TAMBIÉN SE ENCUENTRA AL MENOS EL 75% DE LAS EMPRESAS DENTRO DE UNA MEDIA DE 100 MILLONES, CON OTRAS POCAS QUE SOBREPASAN EN GRAN MEDIDA A LAS DEMÁS, COMO LO ES LA INDUSTRIA DE MANUFACTURING, FINANCE AND INSURANCE, RETAIL TRADE, PROFESSIONAL, SCIENTIFIC AND TECHNICAL SERVICES, INFORMATION Y ARTS, ENTERTAINMENT AND RECREATION.

## For each industry (and for all industries), what can you say about profit margin of firms? show a) descriptive statistics of profit margin and b) plot(s) to illustrate how profit margin changes across industries.

In [ ]:
px.box(last_quarter, x = 'Sector NAICS\nlevel 1', y = 'Operating Profit Margin (OPM)')

SE PUEDE VER QUE LA GRAN MAYORÍA DE LAS INDUSTRIAS CUENTAN CON UN OPM MUY BAJO, CERCANO AL CERO. AUNQUE CON CIERTAS EMPRESAS, PARTICULARMENTE EN LA INDUSTRIA DE MANUFACTURING, DONDE ES INUSUALMENTE BAJO.

In [ ]:
last_quarter['Operating Profit Margin (OPM)'].describe()

count    3354.000000
mean             NaN
std              NaN
min             -inf
25%        -0.258106
50%         0.068118
75%         0.211226
max              inf
Name: Operating Profit Margin (OPM), dtype: float64

DADO QUE PARA OBTENER EL OPM ES NECESARIO DIVIDIR ENTRE EL REVENUE, HAY OCASIONES EN LAS QUE ESTE VALOR ES 0, DÁNDONOS UN RESULTADO INDEFINIDO, POR LO QUE ALGUNAS MEDIDAS ESTADÍSTICAS SE VEN AFECTADAS. SIN EMBARGO, COMO VIMOS ANTERIORMENTE, Y DADA LA DISPERSIÓN DE LOS DATOS, LO MEJOR ES UTILIZAR LA MEDIANA PARA NUESTROS ANÁLISIS, LA CUAL ES 0.068118.

In [ ]:
typical_firm

Sector NAICS\nlevel 1          Book  \
0                                                   -  5.704446e+06   
1                     Accommodation and Food Services  5.139213e+05   
2   Administrative and Support and Waste Managemen...  2.819477e+06   
3          Agriculture, Forestry, Fishing and Hunting  3.629560e+06   
4                 Arts, Entertainment, and Recreation  5.394410e+06   
5                                        Construction  2.535521e+06   
6                                Educational Services  8.931767e+05   
7                               Finance and Insurance  5.482677e+06   
8                   Health Care and Social Assistance  1.080145e+06   
9                                         Information  4.213150e+06   
10                                      Manufacturing  2.417298e+06   
11      Mining, Quarrying, and Oil and Gas Extraction  3.345942e+06   
12      Other Services (except Public Administration)  5.954443e+05   
13   Professional, Scientific, and Technical Services  3.566048e+06   
14                 Real Estate and Rental and Leasing  1.816059e+06   
15                                       Retail Trade  3.379848e+06   
16                     Transportation and Warehousing  3.536723e+06   
17                                          Utilities  7.148295e+06   
18                                    Wholesale Trade  1.529928e+06   

    Book Median   Market Mean  Market Median  OPM Mean  OPM Median  Firms  
0   5704446.000  4.865183e+06   4.865183e+06  0.463653    0.463653      2  
1    243717.500  8.681070e+06   1.394617e+06  0.103713    0.114633     69  
2    566167.000  1.385569e+07   1.938284e+06 -0.109505    0.076840    133  
3   1104345.000  8.046780e+06   1.264045e+06 -0.016727    0.055316     16  
4     67242.769  1.278260e+07   2.504698e+06  0.038509    0.152064     21  
5    998146.500  3.857422e+06   1.745045e+06  0.106393    0.100588     45  
6    649699.000  1.302581e+06   1.524843e+06 -0.001653    0.039924     14  
7   1049158.000  8.412277e+06   1.264517e+06       NaN    0.565211    701  
8    451385.000  3.507730e+06   1.338427e+06       NaN   -0.031415     64  
9    500953.500  1.918280e+07   2.586582e+06       NaN   -0.044653    261  
10   244642.000  1.040643e+07   5.911289e+05       NaN   -0.064151   1565  
11   741145.000  6.783299e+06   1.042405e+06       NaN    0.291527    103  
12   431667.000  2.136156e+06   8.540956e+05  0.043354    0.059704     16  
13   279188.000  1.723014e+07   9.419611e+05       NaN    0.005180    145  
14   634398.000  3.625235e+06   8.819927e+05       NaN    0.145117     68  
15   577426.500  1.998810e+07   1.444971e+06 -0.006559    0.043393    152  
16  1288121.500  1.191248e+07   2.464494e+06  0.010427    0.099848     69  
17  2791950.000  1.588301e+07   6.040649e+06       NaN    0.162881     77  
18   683900.000  4.871519e+06   1.424602e+06  0.013094    0.064806     79

In [ ]:
px.histogram(typical_firm, x = 'Sector NAICS\nlevel 1', y = 'OPM Median')

RETOMANDO LA MEDIANA COMO MEDIDA DE ANÁLISIS, DONDE TENEMOS INFORAMACIÓN DE TODAS LAS INDUSTRIAS, VERMOS QUE EL 68% DE LAS INDUSTRIAS SE ENCUENTRAN POR DEBAJO DEL 0.2, AL 15% POR ENCIEMA DEL 0.2, Y EL RESTO EN VALORES NEGATIVOS. DESTACANDO FINANCE AND INSURANCE CON EL OPM MÁS ALTO.

## Which are the biggest 10 US firms in terms of market value and how far they are from the typical size of a US firm?

In [ ]:
last_quarter[['firm', 'Market','Sector NAICS\nlevel 1']].sort_values('Market', ascending=False).head(10)

firm        Market                              Sector NAICS\nlevel 1
809      AAPL  2.212838e+09                                      Manufacturing
191175   MSFT  1.920840e+09                                        Information
125851  GOOGL  1.435042e+09   Professional, Scientific, and Technical Services
18173    AMZN  1.080624e+09                                       Retail Trade
289525   TSLA  6.976698e+08                                      Manufacturing
296815    UNH  4.818731e+08                              Finance and Insurance
156887    JNJ  4.671001e+08                                      Manufacturing
182535   META  4.363946e+08   Professional, Scientific, and Technical Services
205565   NVDA  3.789750e+08                                      Manufacturing
300325      V  3.743426e+08  Administrative and Support and Waste Managemen...

SE OBTUVIERON LAS 10 EMPRESAS CON MAYOR VALOR DE MERCADO EN ORDEN DESCENDENTE, CABE REALTAR QUE PRÁCTICAMENTE LA MITAD DE ELLAS SON DE LA INDUSTRIA MANUFACTURING, QUE JUNTO CON LAS DEMÁS INDUSTRIAS RECORDEMOS QUE ERAN LAS QUE TENÍAN LOS PICOS MÁS ALTOS EN GRÁFICOS ANTERIORES, LO QUE NOS LLEVÓ A UTILIZAR LA MEDIANA PARA UN MEJOR ANÁLISIS, DADO QUE LA MEDIA YA SOBREPASABA EL VALOR TÍPICO.

In [ ]:
print(last_quarter['Market'].median())

1105075.91083


TOMANDO LA MEDIANA DE 1,105,075.91083, SE PUEDE VER QUE ESTAS EMPRESAS GIGANTES ESTÁN YA MUY POR ENCIMA DEL VALOR TÍPICO, SIMPLEMENTE LA DECIMA EMPRESA ES 338.74 VECES MÁS GRANDE QUE LA MEDIANA DEL MERCADO.

In [ ]:
px.histogram(last_quarter['Market'])

DE MANERA VISUAL AQUÍ SE APRECIA COMO SE ENCUENTRA LA DISTRIBUCIÓN DE MARKET, CON UN SESGO HACIA LA IZQUERDA, DONDE LA GRAN MAYORÍA DE LAS EMPRESAS CUENTAN CON UN VALOR DE APROXIMADAMENTE 5 MILLONES EN MARKET, MIENTRAS QUE DEL OTRO EXTREMO TENEMOS NUESTRAS EMPRESAS MÁS GRANDES, A PENAS VISIBLES.

## Which are the biggest 10 US firms in terms of book value and how far they are from the typical size of a US firm?

In [ ]:
last_quarter[['firm', 'Book','Sector NAICS\nlevel 1']].sort_values('Book', ascending=False).head(10)

firm         Book                             Sector NAICS\nlevel 1
157427    JPM  286143000.0                             Finance and Insurance
34013     BAC  269118000.0                             Finance and Insurance
125851  GOOGL  255419000.0  Professional, Scientific, and Technical Services
49680       C  199626000.0                             Finance and Insurance
319405    XOM  184508000.0                                     Manufacturing
312205    WFC  179793000.0                             Finance and Insurance
191175   MSFT  166542000.0                                       Information
78332     CVX  154562000.0                                     Manufacturing
277650      T  135307000.0                                       Information
18173    AMZN  131402000.0                                      Retail Trade

NUEVAMENTE SE OBTUVIERON LAS EMPRESAS EN ORDEN DESCENDENTE, CABE DESTACAR QUE 3 EMPRESAS QUE  ERAN DE LAS 10 MÁS GRANDES EN MARKET LO SON IGUALMENTE EN BOOK, COMO LO ES GOOGL, MSFT Y AMXN. EN ESTE CASO, LA MAYORÍA DE LAS INDUSTRIAS AHORA SON DE FINANCE AND INSURANCE.

In [ ]:
print(last_quarter['Book'].median())

457737.0


RECORDEMOS QUE DADOS LOS PICOS TAN GRANDES DE VALORES, OPTAMOS POR UTILIZAR LA MEDIANA COMO VALOR TÍPICO, QUE ES 457,737. EVIDENTEMENTE LAS EMPRESAS MÁS GRANDES LO SOBREPASAN POR MUCHO, SIENDO LA ÚLTIMA EMPRESA 287 VECES MÁS GRANDE QUE EL VALOR TÍPICO.

In [ ]:
px.histogram(last_quarter['Book'])

AL IGUAL QUE EN MARKET, SE OBSERVA UN SESGO A LA IZQUIERDA, DONDE LA MAYORÍA DE LAS EMPRESAS TIENEN UN BOOK DE ENTRE 0 Y 2 MILLONES, MIENTRAS QUE EN EL EXTREMOS DEL HISTOGRAMA TENEMOS UNOS CUANTOS PUNTOS, QUE SON NUESTRAS EMPRESAS MÁS GRANDES.

# 2.2.1.2 Considering the whole history of financial data for all firms

## How can you measure firm profitability that can be used to compare performance among firms of different sizes? Select and justify at least 3 measures and show descriptive statistics.

In [94]:
df_history = df_copy.copy()

In [79]:
# Correlación con revenue
df_corrs = df_history.corr()["revenue"]
df_corrs.sort_values(ascending=False)

revenue                          1.000000
cogs                             0.931408
sgae                             0.677615
Market                           0.615881
stockholderequity                0.595631
Book                             0.595626
EBIT                             0.589684
sharesoutstanding                0.494759
Net Income                       0.489768
incometax                        0.449182
finexp                           0.379055
longdebt                         0.378894
otheropexp                       0.333943
totalassets                      0.323492
shortdebt                        0.291061
totalliabilities                 0.275417
originalprice                    0.151463
r2                               0.030197
r                                0.030197
Operating Profit Margin (OPM)    0.005302
Book-to-Market ratio             0.005015
Earnings Per Share (EPS)         0.002891
Sales Anual Growth              -0.001100
adjprice                        -0

In [66]:
df_history[['cogs', 'sgae', 'Market']].describe()

cogs          sgae        Market
count  1.578270e+05  1.578270e+05  1.534490e+05
mean   8.033615e+05  4.027368e+05  1.015524e+07
std    3.753014e+06  1.481353e+06  4.537656e+07
min   -4.460800e+06 -3.373000e+06  9.322199e-01
25%    2.220000e+02  9.277000e+03  3.379509e+05
50%    4.346900e+04  5.600200e+04  1.570573e+06
75%    3.715230e+05  2.167845e+05  5.477976e+06
max    1.155220e+08  5.251500e+07  2.913284e+09

CON EL FIN DE EVALUAR LA RENTABILIDAD, SE BUSCÓ LA CORRELACIÓN DEL REVENUE CON EL RESTO DE LAS VARIABLES, SIENDO LAS 3 QUE ESTÁN MÁS RELACIONADAS EL COGS, EL SGAE Y EL MARKET, CON UN 93.14%, UN 67.76% Y UN 61.58% DE CORRELACIÓN RESPECTIVAMENTE. POR LO QUE ES MUCHO MÁS PROBABLE MEDIR Y PREDECIR LA RENTABILIDAD EN BASE A ESTAS VARIABLES.

DE IGUAL MANERA, 2 DE NUESTRAS 3 VARIABLES SON LA BASE PARA LOS EARNINGS PER SHARE, QUE ES, EFECTIVAMENTE, UNA MÉTRICA PARA MEDIR LA RENTABILIDAD.

## Calculate and explain earnings per share deflated by price.

AL INICIO DE NUESTRO NOTEBOOK SE REALIZÓ EL CÁLCULO DEL EPSP, A CONTINUACIÓN SE MUESTRA NUEVAMENTE EL CÓDIGO COMENTADO PARA OBSERVAR SU CÁLCULO.

In [ ]:
# df_copy['EBIT'] = df_copy['revenue'] - df_copy['cogs'] - df_copy['sgae'] - df_copy['otheropexp']
# df_copy['Net Income'] = df_copy['EBIT'] + df_copy['extraincome']- df_copy['incometax'] - df_copy['finexp']
# df_copy['Earnings Per Share (EPS)'] = df_copy['Net Income'] / df_copy['sharesoutstanding']

# df_copy['EPSP (deflated by price)'] = df_copy['Earnings Per Share (EPS)'] / df_copy['originalprice']

A MODO DE FÓRMULAS SE REQUIRIERON 3 PARA LLEGAR AL EPSP, LAS CUALES SON:

* EBIT = revenue - cogs- sgae - otheropexp

* Net Income = EBIT + extraincome - incometax - finexp

* EPS = Net Income / sharesoutstanding

* EPSP = EPS / originalprice

EL EPS ES UNA FORMA DE MEDIR LA RENTABILIDAD O VALOR DE UN CORPORATIVO, INDICA CUANTO DINERO PUEDE HACER LA EMPRESA POR CADA ACCIÓN.

EL EPSP ES LA TASA COMPARATIVA CON EL PRECIO ORIGINAL DE LA ACCIÓN.

In [77]:
df_history[['firm', 'q', 'EPSP (deflated by price)']]

firm       q  EPSP (deflated by price)
0          A  2000q1                       NaN
1          A  2000q2                  0.004977
2          A  2000q3                  0.006992
3          A  2000q4                  0.012207
4          A  2001q1                  0.010971
...      ...     ...                       ...
323811  ZYNE  2021q2                 -0.045596
323812  ZYNE  2021q3                 -0.060502
323813  ZYNE  2021q4                 -0.074259
323814  ZYNE  2022q1                 -0.097575
323815  ZYNE  2022q2                 -0.198252

[323816 rows x 3 columns]

NUESTROS DATOS Y SU RESPECTIVO EPSP.

# 2.2.2 About statistical modeling

## You have to select a group of firms according to their general industry classification:

* Manufacturing industries

* Commercial industries (retail and wholesale)

* Service industries

* Financial services

In [95]:
df_history

firm       q    revenue       cogs          sgae  otheropexp  \
0          A  2000q1        NaN        NaN           NaN         NaN   
1          A  2000q2  2485000.0  1261000.0  1.010000e+06         0.0   
2          A  2000q3  2670000.0  1369000.0  1.091000e+06         0.0   
3          A  2000q4  3372000.0  1732000.0  1.182000e+06         0.0   
4          A  2001q1  2841000.0  1449000.0  1.113000e+06         0.0   
...      ...     ...        ...        ...           ...         ...   
323811  ZYNE  2021q2        0.0        0.0  9.838494e+03         0.0   
323812  ZYNE  2021q3        0.0        0.0  1.021065e+04         0.0   
323813  ZYNE  2021q4        0.0        0.0  8.836436e+03         0.0   
323814  ZYNE  2022q1        0.0        0.0  8.903915e+03         0.0   
323815  ZYNE  2022q2        0.0        0.0  9.168770e+03         0.0   

         extraincome  finexp  incometax  totalassets  ...  partind  \
0                NaN     NaN        NaN          NaN  ...    0.124   
1       42000.000000   0.000    90000.0  7321000.000  ...    0.124   
2       28000.000000   0.000    83000.0  7827000.000  ...    0.124   
3       10000.000000   0.000   163000.0  8425000.000  ...    0.124   
4       -6000.000000   0.000   119000.0  9208000.000  ...    0.124   
...              ...     ...        ...          ...  ...      ...   
323811   -117.528220  -5.943        0.0    98195.904  ...        -   
323812   -376.636750  -5.038        0.0    89996.170  ...        -   
323813     16.937906  -4.433        0.0    81171.507  ...        -   
323814    317.252110 -96.044        0.0    74381.029  ...        -   
323815   -775.927860 -91.691        0.0    67006.959  ...        -   

               Book        Market         EBIT     Net Income  \
0               NaN  4.700800e+07          NaN            NaN   
1       4642000.000  3.335506e+07  214000.0000  166000.000000   
2       4902000.000  2.216940e+07  210000.0000  155000.000000   
3       5265000.000  2.498606e+07  458000.0000  305000.000000   
4       5541000.000  1.403653e+07  279000.0000  154000.000000   
...             ...           ...          ...            ...   
323811    87699.257  2.182206e+05   -9838.4941   -9950.079320   
323812    78686.999  1.749065e+05  -10210.6520  -10582.250750   
323813    70913.334  1.187065e+05   -8836.4365   -8815.065594   
323814    65166.970  8.701643e+04   -8903.9150   -8490.618890   
323815    57241.873  4.969939e+04   -9168.7695   -9853.006360   

        Earnings Per Share (EPS)  EPSP (deflated by price)  \
0                            NaN                       NaN   
1                       0.367036                  0.004977   
2                       0.342152                  0.006992   
3                       0.668323                  0.012207   
4                       0.337150                  0.010971   
...                          ...                       ...   
323811                 -0.241205                 -0.045596   
323812                 -0.256530                 -0.060502   
323813                 -0.213867                 -0.074259   
323814                 -0.200029                 -0.097575   
323815                 -0.226007                 -0.198252   

        Sales Anual Growth  Operating Profit Margin (OPM)  \
0                      NaN                            NaN   
1                      NaN                       0.086117   
2                      NaN                       0.078652   
3                      NaN                       0.135824   
4                      NaN                       0.098205   
...                    ...                            ...   
323811                 NaN                           -inf   
323812                 NaN                           -inf   
323813                 NaN                           -inf   
323814                 NaN                           -inf   
323815                 NaN                           -inf   

        Book-to-Market ratio  
0           

In [96]:
df_history.replace([np.inf, -np.inf], np.nan, inplace=True)
df_history.dropna(subset=["Market"], how="all", inplace=True)

In [97]:
def dense_inclusive_pct(x):
    r = x.rank(method='dense') - 1
    return r / r.max() * 100

df_history["pct"]=df_history.groupby('q')['Market'].apply(dense_inclusive_pct).astype(int)

df_history["isSmall"] = df_history.pct <= 33
df_history["isSmall"] = df_history["isSmall"].astype(int)

df_history["isMedium"] = (df_history.pct <= 66) & (df_history.pct > 33) 
df_history["isMedium"] = df_history["isMedium"].astype(int)

CLASIFICACIÓN DE EMPRESAS POR TAMAÑO SEGÚN EL PERCENTIL Y EL TRIMESTRE.

In [98]:
df_history

firm       q    revenue       cogs          sgae  otheropexp  \
0          A  2000q1        NaN        NaN           NaN         NaN   
1          A  2000q2  2485000.0  1261000.0  1.010000e+06         0.0   
2          A  2000q3  2670000.0  1369000.0  1.091000e+06         0.0   
3          A  2000q4  3372000.0  1732000.0  1.182000e+06         0.0   
4          A  2001q1  2841000.0  1449000.0  1.113000e+06         0.0   
...      ...     ...        ...        ...           ...         ...   
323811  ZYNE  2021q2        0.0        0.0  9.838494e+03         0.0   
323812  ZYNE  2021q3        0.0        0.0  1.021065e+04         0.0   
323813  ZYNE  2021q4        0.0        0.0  8.836436e+03         0.0   
323814  ZYNE  2022q1        0.0        0.0  8.903915e+03         0.0   
323815  ZYNE  2022q2        0.0        0.0  9.168770e+03         0.0   

         extraincome  finexp  incometax  totalassets  ...         EBIT  \
0                NaN     NaN        NaN          NaN  ...          NaN   
1       42000.000000   0.000    90000.0  7321000.000  ...  214000.0000   
2       28000.000000   0.000    83000.0  7827000.000  ...  210000.0000   
3       10000.000000   0.000   163000.0  8425000.000  ...  458000.0000   
4       -6000.000000   0.000   119000.0  9208000.000  ...  279000.0000   
...              ...     ...        ...          ...  ...          ...   
323811   -117.528220  -5.943        0.0    98195.904  ...   -9838.4941   
323812   -376.636750  -5.038        0.0    89996.170  ...  -10210.6520   
323813     16.937906  -4.433        0.0    81171.507  ...   -8836.4365   
323814    317.252110 -96.044        0.0    74381.029  ...   -8903.9150   
323815   -775.927860 -91.691        0.0    67006.959  ...   -9168.7695   

           Net Income  Earnings Per Share (EPS)  EPSP (deflated by price)  \
0                 NaN                       NaN                       NaN   
1       166000.000000                  0.367036                  0.004977   
2       155000.000000                  0.342152                  0.006992   
3       305000.000000                  0.668323                  0.012207   
4       154000.000000                  0.337150                  0.010971   
...               ...                       ...                       ...   
323811   -9950.079320                 -0.241205                 -0.045596   
323812  -10582.250750                 -0.256530                 -0.060502   
323813   -8815.065594                 -0.213867                 -0.074259   
323814   -8490.618890                 -0.200029                 -0.097575   
323815   -9853.006360                 -0.226007                 -0.198252   

        Sales Anual Growth  Operating Profit Margin (OPM)  \
0                      NaN                            NaN   
1                      NaN                       0.086117   
2                      NaN                       0.078652   
3                      NaN                       0.135824   
4                      NaN                       0.098205   
...                    ...                            ...   
323811                 NaN                            NaN   
323812                 NaN                            NaN   
323813                 NaN                            NaN   
323814                 NaN                            NaN   
323815                 NaN                            NaN   

        Book-to-Market ratio  pct  isSmall  isMedium  
0                        NaN   93        0         0  
1                   0.139169   91        0         0  
2                   0.221116   88        0         0  
3                   0.210718   89        0         0  
4                   0.394756   85        0         0  
...                      ...  ...      ...       ...  
323811              0.401883   19        1         0  
323812              0.449880   17        1         0  
323813              0.597384   15        1         0  
323814              0.748904   12        1    

In [99]:
df_services = df_history.copy()

In [100]:
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == '-')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Agriculture, Forestry, Fishing and Hunting')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Arts, Entertainment, and Recreation')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Construction')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Finance and Insurance')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Health Care and Social Assistance')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Information')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Manufacturing')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Mining, Quarrying, and Oil and Gas Extraction')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Real Estate and Rental and Leasing')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Retail Trade')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Transportation and Warehousing')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Utilities')].index)
df_services = df_services.drop(df_services[(df_services['Sector NAICS\nlevel 1'] == 'Wholesale Trade')].index)

ELIMINAMOS LAS INDUSTRIAS QUE NO NECESITAMOS

In [101]:
df_services = df_services.dropna()

LIMPIAMOS LOS DATOS

In [102]:
df_services

firm       q    revenue       cogs     sgae  otheropexp  extraincome  \
1487    ABM  2012q3  1079235.0   971628.0  84434.0         0.0        698.0   
1488    ABM  2012q4  1090001.0   968416.0  84851.0         0.0        973.0   
1489    ABM  2013q1  1182123.0  1067879.0  94938.0         0.0       1195.0   
1490    ABM  2013q2  1173617.0  1048213.0  91787.0         0.0       1133.0   
1491    ABM  2013q3  1216768.0  1095766.0  92304.0         0.0       1596.0   
...     ...     ...        ...        ...      ...         ...          ...   
323542  ZVO  2021q3    62226.0    36936.0  30081.0       300.0        -69.0   
323543  ZVO  2021q4    54762.0    37543.0  26575.0     14335.0         40.0   
323544  ZVO  2022q1    61633.0    39829.0  29036.0         0.0       -127.0   
323545  ZVO  2022q2    51380.0    34995.0  26108.0     -8882.0      -3824.0   
323725   ZY  2022q2     2634.0     9732.0  58638.0     40460.0       -885.0   

        finexp  incometax  totalassets  ...      EBIT  Net Income  \
1487    2407.0     8887.0    1886895.0  ...   23173.0     12577.0   
1488    2317.0     7727.0    1869251.0  ...   36734.0     27663.0   
1489    3310.0     3809.0    2088148.0  ...   19306.0     13382.0   
1490    3033.0    12443.0    2076390.0  ...   33617.0     19274.0   
1491    3335.0    10883.0    2103403.0  ...   28698.0     16076.0   
...        ...        ...          ...  ...       ...         ...   
323542     0.0       59.0     157775.0  ...   -5091.0     -5219.0   
323543     0.0      -47.0     149034.0  ...  -23691.0    -23604.0   
323544     0.0       78.0     148751.0  ...   -7232.0     -7437.0   
323545     0.0        8.0      85433.0  ...    -841.0     -4673.0   
323725  9376.0       11.0     470968.0  ... -106196.0   -116468.0   

        Earnings Per Share (EPS)  EPSP (deflated by price)  \
1487                    0.231925                  0.012252   
1488                    0.508393                  0.025483   
1489                    0.244877                  0.011011   
1490                    0.352117                  0.014366   
1491                    0.291579                  0.010953   
...                          ...                       ...   
323542                 -0.156174                 -0.065345   
323543                 -0.705446                 -0.555469   
323544                 -0.218383                 -0.266288   
323545                 -0.137054                 -0.141279   
323725                 -1.129214                 -0.918060   

        Sales Anual Growth  Operating Profit Margin (OPM)  \
1487              0.002776                       0.021472   
1488              0.008007                       0.033701   
1489              0.100894                       0.016332   
1490              0.110072                       0.028644   
1491              0.127436                       0.023585   
...                    ...                            ...   
323542           -0.390932                      -0.081815   
323543           -0.412065                      -0.432618   
323544           -0.198103                      -0.117340   
323545           -0.257364                      -0.016368   
323725           -0.552573                     -40.317388   

        Book-to-Market ratio  pct  isSmall  isMedium  
1487                0.804143   54        0         1  
1488                0.783393   56        0         1  
1489                0.707164   55        0         1  
1490                0.651999   56        0         1  
1491                0.609291   56        0         1  
...                      ...  ...      ...       ...  
323542              0.536682    8        1         0  
323543              0.470798    5        1         0  
323544              0.403424    4        1         0  
323545              0.199689    7        1         0  
323725              1.865183   20        1         0  

[13425 rows x 40 columns]

In [106]:
(df_services['revenue'] == 0).sum()

0

VERIFICAMOS QUE NO HAYA CEROS EN REVENUE, DE LO CONTRARIO NOS ARROJARÍA UN INDEFINIDO AL CALCULAR EL OPM.

In [108]:
df_services.columns

Index(['firm', 'q', 'revenue', 'cogs', 'sgae', 'otheropexp', 'extraincome',
       'finexp', 'incometax', 'totalassets', 'totalliabilities', 'shortdebt',
       'longdebt', 'stockholderequity', 'adjprice', 'originalprice',
       'sharesoutstanding', 'lprice', 'r', 'r2', 'Ticker', 'Name', 'Class',
       'Sector NAICS\nlevel 1', 'Exchange / Src', 'Sector\nEconomatica',
       'Sector NAICS\nlast available', 'partind', 'Book', 'Market', 'EBIT',
       'Net Income', 'Earnings Per Share (EPS)', 'EPSP (deflated by price)',
       'Sales Anual Growth', 'Operating Profit Margin (OPM)',
       'Book-to-Market ratio', 'pct', 'isSmall', 'isMedium'],
      dtype='object')

## Using your subset of firms that belong to your industry, which factors (variables) might be related to annual stock return one quarter in the future? Select at least 3 factors and briefly explain why you think might be related to stock returns.

* Do histograms for each factor/variables and interpret them

* Do plots to visualize the possible relationship each factor might have with the dependent variable.

* Show descriptive statistics of these factors

LAS VARIABLES A UTILIZAR SERÁN EL EPSP, SALES ANUAL GROWTH, OPM Y BOOK TO MARKET RATIO. ESTO DEBIDO A QUE SON BUENOS INDICADORES DEL RENDIMIENTO DE LAS EMPRESAS Y NOS DAN UNA IDEA GENERAL DEL TRAYECTO DE LA EMPRESA.

In [111]:
df_services[['firm', 'q', 'EPSP (deflated by price)', 'Sales Anual Growth', 'Operating Profit Margin (OPM)', 'Book-to-Market ratio']]

firm       q  EPSP (deflated by price)  Sales Anual Growth  \
1487    ABM  2012q3                  0.012252            0.002776   
1488    ABM  2012q4                  0.025483            0.008007   
1489    ABM  2013q1                  0.011011            0.100894   
1490    ABM  2013q2                  0.014366            0.110072   
1491    ABM  2013q3                  0.010953            0.127436   
...     ...     ...                       ...                 ...   
323542  ZVO  2021q3                 -0.065345           -0.390932   
323543  ZVO  2021q4                 -0.555469           -0.412065   
323544  ZVO  2022q1                 -0.266288           -0.198103   
323545  ZVO  2022q2                 -0.141279           -0.257364   
323725   ZY  2022q2                 -0.918060           -0.552573   

        Operating Profit Margin (OPM)  Book-to-Market ratio  
1487                         0.021472              0.804143  
1488                         0.033701              0.783393  
1489                         0.016332              0.707164  
1490                         0.028644              0.651999  
1491                         0.023585              0.609291  
...                               ...                   ...  
323542                      -0.081815              0.536682  
323543                      -0.432618              0.470798  
323544                      -0.117340              0.403424  
323545                      -0.016368              0.199689  
323725                     -40.317388              1.865183  

[13425 rows x 6 columns]

In [112]:
px.histogram(df_services, x = 'EPSP (deflated by price)')

In [118]:
df_services['EPSP (deflated by price)'].describe()

count    13425.000000
mean        -0.009644
std          0.181024
min        -14.018716
25%         -0.001442
50%          0.008149
75%          0.013943
max          1.305551
Name: EPSP (deflated by price), dtype: float64

SE PUEDE VER COMO LA GRAN MAYORÍA DE LAS EMPRESAS CUENTAN CON UN EPSP CERCANO A 0, TANTO EN PROMEDIO COMO EN MEDIANA, CON UN MÍNIMO DE HASTA -14. EN GENERAL SE PUEDE VER CIERTA ESTABILIDAD ENTRE LAS EMPRESAS.

In [113]:
px.histogram(df_services, x = 'Sales Anual Growth')

In [119]:
df_services['Sales Anual Growth'].describe()

count     13425.000000
mean         19.274638
std        2170.181229
min        -226.113055
25%          -0.017445
50%           0.070161
75%           0.178606
max      251440.000000
Name: Sales Anual Growth, dtype: float64

EN GENERAL PODEMOS VER UNA INCLINACIÓN HACIA LA IZQUIERDA, NUEVAMENTE LA GRAN MAYORÍA DE LOS DATOS SON CERCANOS AL 0, SIN EMBARGO, HAY UN OUTLIER MUY EXTREMO QUE HACE QUE NUESTRA MEDIA SE VEA AFECTADA, POR LO QUE LO IDEAL ES TOMAR LA MEDIANA.

In [117]:
px.histogram(df_services, x = 'Operating Profit Margin (OPM)')

In [120]:
df_services['Operating Profit Margin (OPM)'].describe()

count    13425.000000
mean        -3.521595
std        222.073959
min     -24214.000000
25%          0.011204
50%          0.078141
75%          0.171453
max        129.649254
Name: Operating Profit Margin (OPM), dtype: float64

AL IGUAL QUE CON LAS VARIABLES ANTERIORES, LOS DATOS SE ENCUENTRAN CERCANOS AL CERO, ESTA VEZ CON UNA INCLINACIÓN HACIA LA DERECHA, YA QUE HAY UN MÍNIMO QUE HACE OUTLIER. LO IDEAL ES CONSIDERAR LA MEDIANA.

In [115]:
px.histogram(df_services, x = 'Book-to-Market ratio')

In [121]:
df_services['Book-to-Market ratio'].describe()

count    13425.000000
mean         0.385184
std          0.684733
min        -11.469108
25%          0.144169
50%          0.305645
75%          0.552295
max         26.225984
Name: Book-to-Market ratio, dtype: float64

AQUÍ SE PUEDE APRECIAR UNA MEJOR DISTRIBUCIÓN, CON UNA CAMPANA DE GAUSS EN LOS DATOS Y UNA MENOR DISPERSIÓN. SI BIEN HAY CIERTOS OUTLIERS, ESTOS NO AFECTAN EN GRAN MEDIDA LOS DATOS, LOS CUALES TAMBIÉN SE ENCUENTRAN CERCANOS AL 0. TANTO LA MEDIA COMO LA MEDIANA SON IDEALES.

In [122]:
df_services['R'] = np.log(df_services.groupby(['firm'])['adjprice'].shift(-1)) - np.log(df_services.groupby(['firm'])['adjprice'].shift(3))

In [123]:
df_services

firm       q    revenue       cogs     sgae  otheropexp  extraincome  \
1487    ABM  2012q3  1079235.0   971628.0  84434.0         0.0        698.0   
1488    ABM  2012q4  1090001.0   968416.0  84851.0         0.0        973.0   
1489    ABM  2013q1  1182123.0  1067879.0  94938.0         0.0       1195.0   
1490    ABM  2013q2  1173617.0  1048213.0  91787.0         0.0       1133.0   
1491    ABM  2013q3  1216768.0  1095766.0  92304.0         0.0       1596.0   
...     ...     ...        ...        ...      ...         ...          ...   
323542  ZVO  2021q3    62226.0    36936.0  30081.0       300.0        -69.0   
323543  ZVO  2021q4    54762.0    37543.0  26575.0     14335.0         40.0   
323544  ZVO  2022q1    61633.0    39829.0  29036.0         0.0       -127.0   
323545  ZVO  2022q2    51380.0    34995.0  26108.0     -8882.0      -3824.0   
323725   ZY  2022q2     2634.0     9732.0  58638.0     40460.0       -885.0   

        finexp  incometax  totalassets  ...  Net Income  \
1487    2407.0     8887.0    1886895.0  ...     12577.0   
1488    2317.0     7727.0    1869251.0  ...     27663.0   
1489    3310.0     3809.0    2088148.0  ...     13382.0   
1490    3033.0    12443.0    2076390.0  ...     19274.0   
1491    3335.0    10883.0    2103403.0  ...     16076.0   
...        ...        ...          ...  ...         ...   
323542     0.0       59.0     157775.0  ...     -5219.0   
323543     0.0      -47.0     149034.0  ...    -23604.0   
323544     0.0       78.0     148751.0  ...     -7437.0   
323545     0.0        8.0      85433.0  ...     -4673.0   
323725  9376.0       11.0     470968.0  ...   -116468.0   

        Earnings Per Share (EPS)  EPSP (deflated by price)  \
1487                    0.231925                  0.012252   
1488                    0.508393                  0.025483   
1489                    0.244877                  0.011011   
1490                    0.352117                  0.014366   
1491                    0.291579                  0.010953   
...                          ...                       ...   
323542                 -0.156174                 -0.065345   
323543                 -0.705446                 -0.555469   
323544                 -0.218383                 -0.266288   
323545                 -0.137054                 -0.141279   
323725                 -1.129214                 -0.918060   

        Sales Anual Growth  Operating Profit Margin (OPM)  \
1487              0.002776                       0.021472   
1488              0.008007                       0.033701   
1489              0.100894                       0.016332   
1490              0.110072                       0.028644   
1491              0.127436                       0.023585   
...                    ...                            ...   
323542           -0.390932                      -0.081815   
323543           -0.412065                      -0.432618   
323544           -0.198103                      -0.117340   
323545           -0.257364                      -0.016368   
323725           -0.552573                     -40.317388   

        Book-to-Market ratio  pct  isSmall  isMedium         R  
1487                0.804143   54        0         1       NaN  
1488                0.783393   56        0         1       NaN  
1489                0.707164   55        0         1       NaN  
1490                0.651999   56        0         1  0.368942  
1491                0.609291   56        0         1  0.383887  
...                      ...  ...      ...       ...       ...  
323542              0.536682    8        1         0 -1.317020  
323543              0.470798    5        1         0 -1.599512  
323544              0.403424    4        1         0 -0.982014  
323545              0.199689    7        1         0       NaN  
323725              1.865183   20        1         0       NaN  

[13425 rows x 41 columns]

In [124]:
df_services = df_services.dropna()

In [125]:
px.histogram(df_services, x = 'R')

In [126]:
df_services['R'].describe()

count    12128.000000
mean         0.064676
std          0.499054
min         -4.826903
25%         -0.145884
50%          0.103483
75%          0.320667
max          3.933887
Name: R, dtype: float64

POR PARTE DE LOS RETURNS, CONTAMOS IGUALMENTE CON UNA DISTRIBUCIÓN NORMAL, CON VALORES CERCANOS AL CERO AUNQUE CON UNA DISPERSIÓN LIGERAMENTE MAYOR.

AHORA EVALUAREMOS LOS DATOS ESTADÍSTICOS GENERALES DE NUESTRAS VARIABLES EN BASE A LOS SERVICIOS.

In [128]:
df_services_stats = df_services.groupby('Sector NAICS\nlevel 1')['EPSP (deflated by price)'].mean().to_frame()
df_services_stats['EPSP median'] = df_services.groupby('Sector NAICS\nlevel 1')['EPSP (deflated by price)'].median()

df_services_stats['Sales Anual Growth mean'] = df_services.groupby('Sector NAICS\nlevel 1')['Sales Anual Growth'].mean()
df_services_stats['Sales Anual Growth median'] = df_services.groupby('Sector NAICS\nlevel 1')['Sales Anual Growth'].median()

df_services_stats['OPM mean'] = df_services.groupby('Sector NAICS\nlevel 1')['Operating Profit Margin (OPM)'].mean()
df_services_stats['OPM median'] = df_services.groupby('Sector NAICS\nlevel 1')['Operating Profit Margin (OPM)'].median()

df_services_stats['Book-to-Market ratio mean'] = df_services.groupby('Sector NAICS\nlevel 1')['Book-to-Market ratio'].mean()
df_services_stats['Book-to-Market ratio median'] = df_services.groupby('Sector NAICS\nlevel 1')['Book-to-Market ratio'].median()

df_services_stats['R mean'] = df_services.groupby('Sector NAICS\nlevel 1')['R'].mean()
df_services_stats['R median'] = df_services.groupby('Sector NAICS\nlevel 1')['R'].median()

df_services_stats.reset_index(inplace=True)
df_services_stats.head()

Sector NAICS\nlevel 1  \
0                    Accommodation and Food Services   
1  Administrative and Support and Waste Managemen...   
2                               Educational Services   
3      Other Services (except Public Administration)   
4   Professional, Scientific, and Technical Services   

   EPSP (deflated by price)  EPSP median  Sales Anual Growth mean  \
0                  0.000458     0.009609               105.119243   
1                 -0.001426     0.008706                 0.508368   
2                 -0.004387     0.009860                 0.081897   
3                 -0.027015     0.008759                 0.886428   
4                 -0.020018     0.006170                 0.173871   

   Sales Anual Growth median  OPM mean  OPM median  Book-to-Market ratio mean  \
0                   0.058911 -9.958606    0.090762                   0.360948   
1                   0.073577  0.079955    0.092734                   0.366804   
2                   0.033784  0.083718    0.094463                   0.611897   
3                   0.042822 -1.744113    0.085350                   0.366646   
4                   0.071879 -5.187968    0.062366                   0.372540   

   Book-to-Market ratio median    R mean  R median  
0                     0.247020  0.092210  0.114469  
1                     0.296557  0.095259  0.132971  
2                     0.528013  0.012269  0.019126  
3                     0.390131 -0.032633  0.037652  
4                     0.311786  0.037074  0.083622

COMO SE HABÍA COMENTADO EN LOS DIAGRAMAS ANTERIORES, LO MÁS RECOMENDABLE ES UTILIZAR LA MEDIANA, DE ESTA MANERA LOS OUTLIERS NO COMPROMETERÁN TANTO NUESTRO ANÁLISIS.

AHORA EN LOS SIGUIENTES HISTOGRAMAS SE MOSTRARÁN NUESTRAS VARIABLES SEGÚN EL TIPO DE SERVICIO.

In [130]:
px.histogram(df_services_stats, x = 'Sector NAICS\nlevel 1', y = 'EPSP median')

SE PUEDE VER QUE LA RENTABILIDAD DE LAS INDUSTRIAS DE SERVICIO SE ENCUENTRAN EN GENERAL ESTABLES Y CON POCAS VARIACIONES ENTRE CADA INDUSTRIA, ÚNICMANTE PROFESSIONAL, SCIENTIFIC AND TECHNICAL SERVICES ES MENOR A LAS DEMÁS.

In [131]:
px.histogram(df_services_stats, x = 'Sector NAICS\nlevel 1', y = 'Sales Anual Growth median')

EN CUANTO AL CRECIMIENTO DE VENTAS ANUALES, DESTACA ADMINISTRATIVE AND SUPPORT AND WASTE MANAGEMENT AND REMEDIATION SERVICES, AL IGUAL QUE PROFESSIONAL, SCIENTIFIC AND TECHNICAL SERVICES. CON INCREMENTOS DEL 0.07.

In [132]:
px.histogram(df_services_stats, x = 'Sector NAICS\nlevel 1', y = 'OPM median')

AL IGUAL QUE CON EL EPSP, LA OPM SE ENCUENTRA ESTABLE Y SIMILAR EN LOS SERVICIOS, SITUANDOSE ALREDEDOR DEL 0.09.

In [133]:
px.histogram(df_services_stats, x = 'Sector NAICS\nlevel 1', y = 'Book-to-Market ratio median')

ES EN EL BOOK-TO-MARKET RATIO DONDE HAY MÁS VARIACIÓN EN LOS DATOS, SIENDO EDUCATIONAL SERVICES EL DE MÁS ALTO RATIO, LO CUAL LO HACE VER COMO UN SECTOR PROMETEDOR.

In [134]:
px.histogram(df_services_stats, x = 'Sector NAICS\nlevel 1', y = 'R median')

EN CUANTO A LOS RETURNS TENEMOS UN ALTO ÍNDICE CON ADMINISTRATIVE AND SUPPORT AND WASTE MANAGEMENT AND REMEDIATION SERVICES, ASÍ COMO EN ACCOMODATION AND FOOD SERVICES.

## Design and run a multiple regression model to examine whether your selected factors and earnings per share deflated by price can explain/predict annual stock returns. You have to control for industry and firm size. To control for these variables you have to include them as extra independent variables in the model

* Your independent variables must be in the right scale so that you can compare the values of the variables among different firms of any size

* For each independent variable you have to check for outliers and do the corresponding adjustments to avoid unreliable results in your regression model

* You must check for possible multicollinearity problems. Briefly explain what is multicollinearity, run and interpret the corresponding test

In [157]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

In [158]:
df_services.columns

Index(['firm', 'q', 'revenue', 'cogs', 'sgae', 'otheropexp', 'extraincome',
       'finexp', 'incometax', 'totalassets', 'totalliabilities', 'shortdebt',
       'longdebt', 'stockholderequity', 'adjprice', 'originalprice',
       'sharesoutstanding', 'lprice', 'r', 'r2', 'Ticker', 'Name', 'Class',
       'Sector NAICS\nlevel 1', 'Exchange / Src', 'Sector\nEconomatica',
       'Sector NAICS\nlast available', 'partind', 'Book', 'Market', 'EBIT',
       'Net Income', 'Earnings Per Share (EPS)', 'EPSP (deflated by price)',
       'Sales Anual Growth', 'Operating Profit Margin (OPM)',
       'Book-to-Market ratio', 'pct', 'isSmall', 'isMedium', 'R'],
      dtype='object')

In [159]:
df_model = df_services[['EPSP (deflated by price)', 'Sales Anual Growth', 'Operating Profit Margin (OPM)', 'Book-to-Market ratio', 'isSmall', 'isMedium', 'R']]

In [160]:
from scipy.stats.mstats import winsorize
df_model["EPSP (deflated by price)"] = winsorize(df_model["EPSP (deflated by price)"], limits=[0.0001, 0.02])
df_model["Sales Anual Growth"] = winsorize(df_model["Sales Anual Growth"], limits=[0.01, 0.0002])
df_model["Operating Profit Margin (OPM)"] = winsorize(df_model["Operating Profit Margin (OPM)"], limits=[0.0001, 0.02])
df_model["Book-to-Market ratio"] = winsorize(df_model["Book-to-Market ratio"], limits=[0.0001, 0.02])
df_model["R"] = winsorize(df_model["R"], limits=[0.0001, 0.02])

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/st

SE REALIZÓ UNA WINZORIZACIÓN PARA AJUSTAR LOS OUTLIERS SUPERIORES E INFERIORES.

In [161]:
df_model.rename(columns = {'EPSP (deflated by price)':'EPSP', 'Sales Anual Growth':'SAG', 'Operating Profit Margin (OPM)':'OPM', 'Book-to-Market ratio':'BtMr'}, inplace = True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:5047: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [162]:
df_model

EPSP       SAG       OPM      BtMr  isSmall  isMedium         R
1490    0.014366  0.110072  0.028644  0.651999        0         1  0.368942
1491    0.010953  0.127436  0.023585  0.609291        0         1  0.383887
1492    0.015225  0.134653  0.030243  0.577989        0         1  0.280456
1493    0.008140  0.037516  0.019513  0.577989        0         1  0.118580
1494    0.010060  0.049150  0.022659  0.622654        0         1 -0.007364
...          ...       ...       ...       ...      ...       ...       ...
323540 -0.071394 -0.214699 -0.121482  0.382846        1         0 -0.166757
323541 -0.046721 -0.334366 -0.064883  0.545731        1         0 -0.515001
323542 -0.065345 -0.390932 -0.081815  0.536682        1         0 -1.317020
323543 -0.555469 -0.412065 -0.432618  0.470798        1         0 -1.599512
323544 -0.266288 -0.198103 -0.117340  0.403424        1         0 -0.982014

[12128 rows x 7 columns]

SE RENOMBRARON LAS COLUMNAS PARA QUE SEA ACEPTADA SU SINTAXIS.

In [163]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
vif_data = pd.DataFrame()
vif_data["feature"] = df_model.columns
  
# calculating VIF for each feature
vif_data["VIF"] = [variance_inflation_factor(df_model.values, i)
                          for i in range(len(df_model.columns))]

print(vif_data)

    feature       VIF
0      EPSP  1.136090
1       SAG  1.001478
2       OPM  1.006944
3      BtMr  1.481623
4   isSmall  1.383170
5  isMedium  1.146429
6         R  1.113232


EL VIF ES MENOR A 1.5, POR LO QUE PODEMOS DECIR QUE NO HAY MULTICOLINEARIDAD EN LOS DATOS DE LAS VARIABLES.

In [164]:
mod = smf.ols('R ~ EPSP + SAG + OPM + BtMr + isSmall + isMedium', data = df_model).fit()

print(mod.summary())

                            OLS Regression Results                            
Dep. Variable:                      R   R-squared:                       0.118
Model:                            OLS   Adj. R-squared:                  0.118
Method:                 Least Squares   F-statistic:                     270.6
Date:                Mon, 12 Sep 2022   Prob (F-statistic):               0.00
Time:                        11:30:27   Log-Likelihood:                -7485.2
No. Observations:               12128   AIC:                         1.498e+04
Df Residuals:                   12121   BIC:                         1.504e+04
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.1449      0.007     20.104      0.0

## Interpret your model

* Interpret the results of each coefficient (beta and their statistical significance)

* Interpret the R-squared of the model

TENIENDO COMO VARIABLE DEPENDENDIENTE NUESTRA R SABEMOS QUE:

* EL EPSP CUENTA CON UNA RELACIÓN POSITIVA DE 1.0318, ESTO QUIERE DECIR QUE POR CADA INCREMENTO EN UNA UNIDAD AL EPSP, LA R AUMENTA EN 1.0318. ESTA VARIABLE ES SIGNIFICATIVA, PUES CUENTA CON UN T VALUE DE 34.302, POR LO QUE NOS DA UN 99.99% DE FIDELIDAD.

* CON SALES ANUAL GROWTH HAY UNA RELACIÓN POSITIVA, DONDE POR CADA CAMBIO EN UNA UNIDAD, EL RETORNO DE STOCK AUMENTA EN 6.234E-5, UN INCREMENTO BASTANTE PEQUEÑO, ADEMÁS DE POCO SIGNIFICATIVO, SU T VALUE ES DE 0.211, ENTONCES TIENE UNA FIDELIDAD MUY BAJA.

* EL OPM TIENE IGUALMENTE UNA RELACIÓN POSITIVA, CON UN INCREMENTO EN R DE A PENAS 0.0001 POR CADA UNIDAD EN OPM, SIN EMBARGO, DADO SU T VALUE DE 2.940, ESTA VARIABLE TIENE UNA FIDELIDAD DEL 99.99%.

* HABLANDO DEL BOOK TO MARKET RATIO, ESTA TIENE UNA RELACIÓN NEGATIVA, DONDE POR CADA CAMBIO EN UNA UNIDAD, EL R SE VERÁ DISMINUIDO EN 0.0731. ES UNA VARIABLE SIGNIFICATIVA EN UN 99.995 DADO SU T VALUE.

* AHORA EN CUANTO AL TAMAÑO DE LAS EMPRESAS SABEMOS QUE:

  * LAS EMPRESAS GRANDES TENDRÁN MAYORES RETORNOS DADO QUE INICIAN CON UN 0.1449, DADA LA INSTERSECCIÓN.
  * LAS MEDIANAS INICIARÁN EN 0.1021 (0.1449 - 0.0428)
  * LAS CHICAS INICIARÁN EN 0.0302 (0.1449 - 0.1147)
  * ESTAS VARIABLES SON SIGNIFICTIVAS POR SUS ALTOS VALORES DE T VALUE, TENIENDO 99.99% DE FIDELIDAD.

* EN GENERAL NUESTRO MODELO SOLO TIENE UNA PRECISIÓN DEL 11.8%, EXPRESADO POR NUESTRA R CUADRADA, ESTO TAMBIÉN SE DEBE A QUE QUEREMOS PREDECIR LOS RETORNOS ANUALES, Y EL MERCADO CUENTA CON LA FACILIDAD DE ADAPTARSE A LOS CAMBIOS, POR LO QUE UNA PREDICCIÓN MÁS SIGNIFICATIVA REQUIERE QUE LA PREDICCIÓN SEA A UN INTÉRVALO DE TIEMPO MENOR.

ISBIG

* R = 0.1449 + 1.0318(EPSP) + 0.0001(OPM) - 0.0731(BOOK-TO-MARKET RATIO)

ISMEDIUM

* R = 0.1021 + 1.0318(EPSP) + 0.0001(OPM) - 0.0731(BOOK-TO-MARKET RATIO)

ISSMALL

* R = 0.0302 + 1.0318(EPSP) + 0.0001(OPM) - 0.0731(BOOK-TO-MARKET RATIO)